### ISSUE TITLE
Migrating media from non-standard folder

### ISSUE THREAD

**Issue by dizzyskips** [deliciousbrains/wp-migrate-db-pro-bugs#4]
Hi Gents,

Just tried out the new beta 1.3b2 of Migrate DB Pro along with the Media Files Add-on Version 1.0b2 and it worked splendidly, but I did find something a bit odd.

I was using Migrate DB Pro to set up a local dev environment by:
1. checking out a repo containing the entire site I'm working on (WP install, plugins, themes, and hopefully uploads).
2. then I installed WP with a blank database
3. activated WP Migrate DB Pro on both sites and added the Media Add-on
4. after adding license keys, I did a pull from the development server to my local dev machine to include media.

All of the database migrated perfectly, like usual, and I saw a bunch of files from the `wp-content/uploads` directory get pulled down too.  So far, so good.

The issue is that we often build on Roots Theme (roots.io) and it defaults to saving uploads in a top-level directory called `/assets`.

![screenshot 2013-12-05 14 30 50](https://f.cloud.github.com/assets/2308362/1687965/3354be5a-5dfe-11e3-9305-e07e9c349b59.png)

After the initial migration, another developer asked me if it caught the files in the /assets directory and, looking through the migrated files, it seemed like it didn't get them all, so I ran a test.

I verified that _Settings > Media_ was set to `/assets` on both my local and staging server installs and then uploaded a new file to the staging server, which was uploaded to the `/assets` folder as expected.

Then, I re-ran the same exact, saved migration profile to see if it would pull down that file and it actually downloaded all of the files in the `/assets` directory, like it missed them the first time.

![screenshot 2013-12-05 14 24 33](https://f.cloud.github.com/assets/2308362/1687999/c7ff996c-5dfe-11e3-8886-53f13531795e.png)

Let me know what other info I can provide to help diagnose this issue.  I'd be happy to provide you with an API Key, if you wish to try the migration yourself.

Cheers!
Eric Amundson

**Comment by aprea** [comment]
Thanks for the detailed report Eric, I'll be looking into this today and will report back with my findings.

**Comment by dizzyskips** [comment]
My pleasure.  Just holler if I can provide anything else. :)

**Comment by aprea** [comment]
I'm having a little difficulty replicating this issue. I don't have the "uploading files" section in my media settings as you do so I instead used the constant:

``` php
define('UPLOADS', 'assets');
```

I believe it does the same thing though.

So far I've tested the following:
- Pull from website using **default** uploads directory into website using **default** uploads directory
- Pull from website using **custom** uploads directory into website using **default** uploads directory
- Pull from website using **default** uploads directory into website using **custom** uploads directory
- Pull from website using **custom** uploads directory into website using **custom** uploads directory

They've all seemed to work for me so far.

A question, you said this:

> I verified that Settings > Media was set to /assets on both my local and staging server installs and then uploaded a new file to the staging server, which was uploaded to the /assets folder as expected.

Did you actually change anything here or were you just verifying the settings?

**Comment by dizzyskips** [comment]
> So far I've tested the following:
> 
> Pull from website using default uploads directory into website using default uploads directory
> Pull from website using custom uploads directory into website using default uploads directory
> Pull from website using default uploads directory into website using custom uploads directory
> Pull from website using custom uploads directory into website using custom uploads directory

Embarrassingly, I didn't know you could do this - I just sort of assumed that the plugin would look at the settings for uploads on the site being migrated and mimic the location when migrating uploads too.

How do you tell it where to migrate the files to, if you're changing the uploads directory?

In this case, it seems like it migrated everything from the default uploads directory on first pass, and actually pulled from the `assets` directory on second migration with no other changes.

> Did you actually change anything here or were you just verifying the settings?

No changes, just verifying settings.

Again, I'm happy to give you WP access and an API key to migrate data from a dev site we're working on, if you think it would help.

**Comment by aprea** [comment]
> How do you tell it where to migrate the files to, if you're changing the uploads directory?

It will always migrate the files to the local site's (the one you're pulling into) defined "uploads" directory. It disregards the remote site's folder structure because it's mostly irrelevant.

Here's one problem I could envision happening when changing around the uploads directory.
- You installed a fresh copy of WordPress
- You upload cat.jpg
- You change your uploads folder to "assets"
- You upload dog.jpg

Now your database has both cat.jpg and dog.jpg but it can only find dog.jpg because you folder structure looks like this

assets
-- dog.jpg
wp-content
-- uploads
---- cat.jpg

Is this something that could have happened?

**Comment by dizzyskips** [comment]
Howdy Chris,

Okay, so I think you've just answered the question for me. I didn't intentionally change it, but I did a default WP install.

To clarify, here's more detail on how I've been setting up my local development environment:
1. check out a git repo containing the entire site to a directory on my laptop's local web server.
2. setup a blank database.
3. setup my local hosts file and apache vhost to point to dev.mydomain.com.
4. using a browser, I hit dev.mydomain.com, and continue with a default install of WordPress.
5. since it's a new, blank database that I'll replace with data migrated using your spiffy plugin, I quickly put in enough data to login.
6. activate WP Migrate DB Pro on both sites and add the Media Add-on.
7. after adding license keys, I did a pull from the server to my laptop's dev install.

Of course, since a default WP install comes with Twenty Thirteen and the uploads directory set to the default location, the first migration is downloading all files to the `wp-content/uploads` folder. 

Then, once the database has been replaced, the setting for media directory location is updated to reflect `assets` as it is on the server.  Then, when I pulled again, it re-downloaded the files to the correct location.

I can simply workaround this by making sure to set the Settings > Media location to `assets` before my first pull.

If you were looking to make the plugin idiot-proof (or at least Eric-proof), maybe the plugin could _read_ the destination's media's storage location ahead of time and alert you if the location is different? 

And/or could you provide an option to manually set the destination for migrated media?

Cheers!

**Comment by aprea** [comment]
Both great suggestions Eric, I really appreciate the time you've put into this issue.

Let me run this by the guys and see what they think.

**Comment by dizzyskips** [comment]
My pleasure.  Thank you guys for such a useful tool and for giving us beta access. :)

The addition of file migration is another nifty time-saver in our dev process. Thanks!

**Comment by bradt** [comment]
We are planning to make a change to the migration process so that the database migration is finalized before starting the media file migration. Correct me if I'm wrong, but wouldn't that fix the issue here?

**Comment by dizzyskips** [comment]
Hey Brad,

Yep, that should definitely fix the issue.  Thanks!

**Comment by aprea** [comment]
Closing this issue for now, we'll be releasing an update for this soon, please feel free to raise another issue if the problem is still relevant.
